src/shared/sam/debugger/goSamDebug.ts

```diff
@@ -0,0 +1,185 @@
+
+    if (!config.codeRoot) {
+        throw Error('missing launch.json, template.yaml, and failed to discover project root')
+    }
+
+    let localRoot: string | undefined
+    let remoteRoot: string | undefined
+
+    config.debugPort = config.debugPort ?? -1
+    config.codeRoot = pathutil.normalize(config.codeRoot)
+    config.debuggerPath = GO_DEBUGGER_PATH
+
+    // Always generate a temporary template.yaml, don't use workspace one directly.
+    config.templatePath = await makeInputTemplate(config)
+
+    const isImageLambda = isImageLambdaConfig(config)
+
+    if (isImageLambda && !config.noDebug) {
+        config.containerEnvVars = {
+            _AWS_LAMBDA_GO_DEBUGGING: '1',
+            _AWS_LAMBDA_GO_DELVE_API_VERSION: '2',
+            _AWS_LAMBDA_GO_DELVE_PATH: path.join(GO_DEBUGGER_PATH, 'dlv'),
+            _AWS_LAMBDA_GO_DELVE_LISTEN_PORT: config.debugPort.toString(),
+        }
+    }
+
+    // if provided, use the user's mapping instead
+    if (config.lambda?.pathMappings !== undefined && config.lambda.pathMappings.length > 0) {
+        const mappings = config.lambda.pathMappings
+        if (mappings.length !== 1) {
+            getLogger().warn(
+                'This language only supports a single path mapping entry. Taking the first entry in the list.'
+            )
+        }
+        localRoot = mappings[0].localRoot
+        remoteRoot = mappings[0].remoteRoot
+    }
+
+    //  Make a go launch-config from the generic config.
+    const goLaunchConfig: GoDebugConfiguration = {
+        ...config, // Compose.
+        type: 'go',
+        processName: 'godelve',
+        request: 'attach',
+        mode: config.noDebug ? undefined : 'remote',
+        runtimeFamily: RuntimeFamily.Go,
+        preLaunchTask: undefined,
+        host: 'localhost',
+        port: config.debugPort,
+        skipFiles: [],
+        debugArgs: isImageLambda ? undefined : ['-delveAPI=2'],
+        localRoot: localRoot ?? config.codeRoot,
+        remoteRoot: remoteRoot ?? '/var/task',
+    }
+
+    return goLaunchConfig
+}
+
+/**
+ * @param debuggerPath Installation path for the debugger
+ * @param isWindows Flag for making a windows script
+ * @param forceDirect Sets GOPROXY to direct to prevent DNS failures, for use in tests *only*. See https://golang.org/ref/mod#module-proxy
+ * @returns Path for the debugger install script
+ */
+async function makeInstallScript(debuggerPath: string, isWindows: boolean, forceDirect: boolean): Promise<string> {
+    let script: string = ''
+    const DELVE_MODULE: string = path.normalize('github.com/go-delve/delve/cmd/dlv')
+    const scriptExt: string = isWindows ? 'cmd' : 'sh'
+    const installScriptPath: string = path.join(debuggerPath, `install.${scriptExt}`)
+    const delvePath: string = path.join(debuggerPath, 'dlv')
+
+    // TODO: don't just check if the file exists, ideally we should check for a version too
+    const alreadyInstalled = await SystemUtilities.fileExists(installScriptPath)
```

Comment:
We should store the version we installed to make this easier to do in the future

Reply:
Will do in the 2nd PR. There should be a way to get the version of the `dlv` package but not sure what `go` command exposes that.